File: src/main/java/io/github/legacy_fabric_community/serialization/nbt/NbtOps.java
package io.github.legacy_fabric_community.serialization.nbt;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map.Entry;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import com.google.common.collect.Iterators;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.google.common.collect.PeekingIterator;
import com.mojang.datafixers.DataFixUtils;
import com.mojang.datafixers.util.Pair;
import com.mojang.serialization.DataResult;
import com.mojang.serialization.DynamicOps;
import com.mojang.serialization.MapLike;
import com.mojang.serialization.RecordBuilder;
import com.mojang.serialization.RecordBuilder.AbstractStringBuilder;
import io.github.legacy_fabric_community.serialization.codec.ExtendedOps;
import io.github.legacy_fabric_community.serialization.mixin.CompoundTagAccessor;
import io.github.legacy_fabric_community.serialization.mixin.EndTagAccessor;
import io.github.legacy_fabric_community.serialization.mixin.ListTagAccessor;
import io.github.legacy_fabric_community.serialization.mixin.TagAccessor;

import net.minecraft.nbt.ByteArrayTag;
import net.minecraft.nbt.ByteTag;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.DoubleTag;
import net.minecraft.nbt.EndTag;
import net.minecraft.nbt.FloatTag;
import net.minecraft.nbt.IntArrayTag;
import net.minecraft.nbt.IntTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.nbt.LongTag;
import net.minecraft.nbt.ShortTag;
import net.minecraft.nbt.StringTag;
import net.minecraft.nbt.Tag;

import net.fabricmc.fabric.api.util.NbtType;

public class NbtOps implements ExtendedOps<Tag> {
    public static final NbtOps INSTANCE = new NbtOps();
    static final EndTag THE_END_TAG = EndTagAccessor.create();

    private NbtOps() {
    }

    private static Tag createListTag(byte b, byte c) {
        if (matches(b, c, (byte) 1)) {
            return new ByteArrayTag(new byte[0]);
        }return matches(b, c, (byte) 3) ? new IntArrayTag(new int[0]) : new ListTag();
    }

    private static boolean matches(byte b, byte c, byte d) {
        return b == d && (c == d || c == 0);
    }

    private static void fill(ListTag listTag, Tag tag, Tag tag2) {
        if (tag instanceof ListTag) {
            ListTag abstractListTag2 = (ListTag) tag;
            ((ListTagAccessor) abstractListTag2).getValue().forEach(listTag::add);
        }
        listTag.add(tag2);
    }

    private static void fillAll(ListTag listTag, Tag tag, List<Tag> list) {
        if (tag instanceof ListTag) {
            ListTag tempList = (ListTag) tag;
            ((ListTagAccessor) tempList).getValue().forEach(listTag::add);
        }
        list.forEach(listTag::add);
    }

    public Tag empty() {
        return THE_END_TAG;
    }

    public <U> U convertTo(DynamicOps<U> dynamicOps, Tag tag) {
        switch (tag.getType()) {
            case NbtType.END:
                return dynamicOps.empty();
            case NbtType.BYTE:
                return dynamicOps.createByte(((Tag.NumberTag) tag).getByte());
            case NbtType.SHORT:
                return dynamicOps.createShort(((Tag.NumberTag) tag).getShort());
            case NbtType.INT:
                return dynamicOps.createInt(((Tag.NumberTag) tag).getInt());
            case NbtType.LONG:
                return dynamicOps.createLong(((Tag.NumberTag) tag).getLong());
            case NbtType.FLOAT:
                return dynamicOps.createFloat(((Tag.NumberTag) tag).getFloat());
            case NbtType.DOUBLE:
                return dynamicOps.createDouble(((Tag.NumberTag) tag).getDouble());
            case NbtType.BYTE_ARRAY:
                return dynamicOps.createByteList(ByteBuffer.wrap(((ByteArrayTag) tag).getArray()));
            case NbtType.STRING:
                return dynamicOps.createString(((TagAccessor) tag).invokeAsString());
            case NbtType.LIST:
                return this.convertList(dynamicOps, tag);
            case NbtType.COMPOUND:
                return this.convertMap(dynamicOps, tag);
            case NbtType.INT_ARRAY:
                return dynamicOps.createIntList(Arrays.stream(((IntArrayTag) tag).getIntArray()));
            default:
                throw new IllegalStateException("Unknown tag type: " + tag);
        }
    }

    public DataResult<Number> getNumberValue(Tag tag) {
        return tag instanceof Tag.NumberTag ? DataResult.success((((NumericTag) tag)).getNumber()) : DataResult.error("Not a number");
    }

    public Tag createNumeric(Number number) {
        if (number instanceof Integer || number instanceof BigInteger || number instanceof AtomicInteger) {
            return this.createInt(number.intValue());
        } else if (number instanceof Float) {
            return this.createFloat(number.floatValue());
        } else if (number instanceof Byte) {
            return this.createByte(number.byteValue());
        } else if (number instanceof Long || number instanceof AtomicLong) {
            return this.createLong(number.longValue());
        } else if (number instanceof Short) {
            return this.createShort(number.shortValue());
        } else if (number instanceof Double || number instanceof BigDecimal) {
            return this.createDouble(number.doubleValue());
        } else {
            throw new UnsupportedOperationException("Only Primitive Numbers are supported!");
        }
    }

    public Tag createByte(byte b) {
        return new ByteTag(b);
    }

    public Tag createShort(short s) {
        return new ShortTag(s);
    }

    public Tag createInt(int i) {
        return new IntTag(i);
    }

    public Tag createLong(long l) {
        return new LongTag(l);
    }

    public Tag createFloat(float f) {
        return new FloatTag(f);
    }

    public Tag createDouble(double d) {
        return new DoubleTag(d);
    }

    public Tag createBoolean(boolean bl) {
        return new ByteTag((byte) (bl ? 1 : 0));
    }

    public DataResult<String> getStringValue(Tag tag) {
        return tag instanceof StringTag ? DataResult.success(((TagAccessor) tag).invokeAsString()) : DataResult.error("Not a string");
    }

    public Tag createString(String string) {
        return new StringTag(string);
    }

    public DataResult<Tag> mergeToList(Tag tag, Tag tag2) {
        if (!(tag instanceof ListTag) && !(tag instanceof EndTag)) {
            return DataResult.error("mergeToList called with not a list: " + tag, tag);
        } else {
            Tag listTag = createListTag((byte) (tag instanceof ListTag ? ((ListTag) tag).getElementType() : 0), tag2.getType());
            if (listTag instanceof ListTag) {
                fill((ListTag) listTag, tag, tag2);
            }
            return DataResult.success(listTag);
        }
    }

    public DataResult<Tag> mergeToList(Tag tag, List<Tag> list) {
        if (!(tag instanceof ListTag) && !(tag instanceof EndTag)) {
            return DataResult.error("mergeToList called with not a list: " + tag, tag);
        } else {
            Tag abstractListTag = createListTag(tag instanceof ListTag ? (byte) ((ListTag) tag).getElementType() : (byte) 0, list.stream().findFirst().map(Tag::getType).orElse((byte) 0));
            if (abstractListTag instanceof ListTag) {
                fillAll((ListTag) abstractListTag, tag, list);
            }
            return DataResult.success(abstractListTag);
        }
    }

    public DataResult<Tag> mergeToMap(Tag mapTag, Tag keyTag, Tag tag) {
        if (!(mapTag instanceof CompoundTag) && !(mapTag instanceof EndTag)) {
            return DataResult.error("mergeToMap called with not a map: " + mapTag, mapTag);
        } else if (!(keyTag instanceof StringTag)) {
            return DataResult.error("key is not a string: " + keyTag, mapTag);
        } else {
            CompoundTag compoundTag = new CompoundTag();
            if (mapTag instanceof CompoundTag) {
                CompoundTag compoundTag2 = (CompoundTag) mapTag;
                compoundTag2.getKeys().forEach((string) -> compoundTag.put(string, compoundTag2.get(string)));
            }

            compoundTag.put(((TagAccessor) keyTag).invokeAsString(), tag);
            return DataResult.success(compoundTag);
        }
    }

    public DataResult<Tag> mergeToMap(Tag tag, MapLike<Tag> mapLike) {
        if (!(tag instanceof CompoundTag) && !(tag instanceof EndTag)) {
            return DataResult.error("mergeToMap called with not a map: " + tag, tag);
        } else {
            CompoundTag compoundTag = new CompoundTag();
            if (tag instanceof CompoundTag) {
                CompoundTag compoundTag2 = (CompoundTag) tag;
                compoundTag2.getKeys().forEach((string) -> compoundTag.put(string, compoundTag2.get(string)));
            }

            List<Tag> list = Lists.newArrayList();
            mapLike.entries().forEach((pair) -> {
                Tag tag2 = pair.getFirst();
                //noinspection ConstantConditions
                if (!(tag instanceof StringTag)) {
                    list.add(tag);
                } else {
                    compoundTag.put(((TagAccessor) tag2).invokeAsString(), pair.getSecond());
                }
            });
            return !list.isEmpty() ? DataResult.error("some keys are not strings: " + list, compoundTag) : DataResult.success(compoundTag);
        }
    }

    public DataResult<Stream<Pair<Tag, Tag>>> getMapValues(Tag tag) {
        if (!(tag instanceof CompoundTag)) {
            return DataResult.error("Not a map: " + tag);
        } else {
            CompoundTag compoundTag = (CompoundTag) tag;
            return DataResult.success(compoundTag.getKeys().stream().map((string) -> Pair.of(this.createString(string), compoundTag.get(string))));
        }
    }

    public DataResult<Consumer<BiConsumer<Tag, Tag>>> getMapEntries(Tag tag) {
        if (!(tag instanceof CompoundTag)) {
            return DataResult.error("Not a map: " + tag);
        } else {
            CompoundTag compoundTag = (CompoundTag) tag;
            return DataResult.success((consumer) -> compoundTag.getKeys().forEach((string) -> consumer.accept(this.createString(string), compoundTag.get(string))));
        }
    }

    public DataResult<MapLike<Tag>> getMap(Tag tag) {
        if (!(tag instanceof CompoundTag)) {
            return DataResult.error("Not a map: " + tag);
        } else {
            final CompoundTag compoundTag = (CompoundTag) tag;
            return DataResult.success(new MapLike<Tag>() {
                public Tag get(Tag tag) {
                    return compoundTag.get(((TagAccessor) tag).invokeAsString());
                }

                public Tag get(String string) {
                    return compoundTag.get(string);
                }

                public Stream<Pair<Tag, Tag>> entries() {
                    return compoundTag.getKeys().stream().map((string) -> Pair.of(NbtOps.this.createString(string), compoundTag.get(string)));
                }

                public String toString() {
                    return "MapLike[" + compoundTag + "]";
                }
            });
        }
    }

    public Tag createMap(Stream<Pair<Tag, Tag>> stream) {
        CompoundTag compoundTag = new CompoundTag();
        stream.forEach((pair) -> compoundTag.put(((TagAccessor) pair.getFirst()).invokeAsString(), pair.getSecond()));
        return compoundTag;
    }

    public DataResult<Stream<Tag>> getStream(Tag tag) {
        return tag instanceof ListTag ? DataResult.success(((ListTagAccessor) tag).getValue().stream()) : DataResult.error("Not a list");
    }

    public DataResult<Consumer<Consumer<Tag>>> getList(Tag tag) {
        if (tag instanceof ListTag) {
            ListTag listTag = (ListTag) tag;
            return DataResult.success(((ListTagAccessor) listTag).getValue()::forEach);
        } else {
            return DataResult.error("Not a list: " + tag);
        }
    }

    public DataResult<ByteBuffer> getByteBuffer(Tag tag) {
        return tag instanceof ByteArrayTag ? DataResult.success(ByteBuffer.wrap(((ByteArrayTag) tag).getArray())) : ExtendedOps.super.getByteBuffer(tag);
    }

    public Tag createByteList(ByteBuffer byteBuffer) {
        return new ByteArrayTag(DataFixUtils.toArray(byteBuffer));
    }

    public DataResult<IntStream> getIntStream(Tag tag) {
        return tag instanceof IntArrayTag ? DataResult.success(Arrays.stream(((IntArrayTag) tag).getIntArray())) : ExtendedOps.super.getIntStream(tag);
    }

    public Tag createIntList(IntStream intStream) {
        return new IntArrayTag(intStream.toArray());
    }

    public Tag createList(Stream<Tag> stream) {
        PeekingIterator<Tag> peekingIterator = Iterators.peekingIterator(stream.iterator());
        if (!peekingIterator.hasNext()) {
            return new ListTag();
        } else {
            Tag tag = peekingIterator.peek();
            ArrayList<Number> list3;
            if (tag instanceof ByteTag) {
                list3 = Lists.newArrayList(Iterators.transform(peekingIterator, (tagx) -> tagx != null ? ((ByteTag) tagx).getByte() : 0));
                byte[] arr = new byte[list3.size()];
                for (int a = 0; a < arr.length; a++) {
                    arr[a] = (byte) list3.get(a);
                }
                return new ByteArrayTag(arr);
            } else if (tag instanceof IntTag) {
                list3 = Lists.newArrayList(Iterators.transform(peekingIterator, (tagx) -> tagx != null ? ((IntTag) tagx).getInt() : 0));
                int[] arr = new int[list3.size()];
                for (int a = 0; a < arr.length; a++) {
                    arr[a] = (int) list3.get(a);
                }
                return new IntArrayTag(arr);
            } else {
                ListTag listTag = new ListTag();

                while (peekingIterator.hasNext()) {
                    Tag tag2 = peekingIterator.next();
                    if (!(tag2 instanceof EndTag)) {
                        listTag.add(tag2);
                    }
                }

                return listTag;
            }
        }
    }

    public Tag remove(Tag tag, String string) {
        if (tag instanceof CompoundTag) {
            CompoundTag compoundTag = (CompoundTag) tag;
            CompoundTag compoundTag2 = new CompoundTag();
            compoundTag.getKeys().stream().filter((string2) -> !Objects.equals(string2, string)).forEach((stringx) -> compoundTag2.put(stringx, compoundTag.get(stringx)));
            return compoundTag2;
        } else {
            return tag;
        }
    }

    public String toString() {
        return "Nbt";
    }

    public RecordBuilder<Tag> mapBuilder() {
        return new NbtOps.MapBuilder();
    }

    class MapBuilder extends AbstractStringBuilder<Tag, CompoundTag> {
        protected MapBuilder() {
            super(NbtOps.this);
        }

        protected CompoundTag initBuilder() {
            return new CompoundTag();
        }

        protected CompoundTag append(String string, Tag tag, CompoundTag compoundTag) {
            compoundTag.put(string, tag);
            return compoundTag;
        }

        protected DataResult<Tag> build(CompoundTag compoundTag, Tag tag) {
            if (tag != null && tag != THE_END_TAG) {
                if (!(tag instanceof CompoundTag)) {
                    return DataResult.error("mergeToMap called with not a map: " + tag, tag);
                } else {
                    CompoundTag compoundTag2 = new CompoundTag();
                    ((CompoundTagAccessor) compoundTag).setData(Maps.newHashMap(((CompoundTagAccessor) tag).getData()));

                    for (Entry<String, Tag> stringTagEntry : ((CompoundTagAccessor) compoundTag).getData().entrySet()) {
                        compoundTag2.put(stringTagEntry.getKey(), (stringTagEntry).getValue());
                    }

                    return DataResult.success(compoundTag2);
                }
            } else {
                return DataResult.success(compoundTag);
            }
        }
    }
}

File: shadowed-config-nbt/src/main/scala/io/github/legacy_fabric_community/shadowedconfig/nbt/NbtOps.scala
package io.github.legacy_fabric_community.shadowedconfig.nbt

import java.math.{BigDecimal, BigInteger}
import java.nio.ByteBuffer
import java.util
import java.util.Objects
import java.util.concurrent.atomic.{AtomicInteger, AtomicLong}
import java.util.function.{BiConsumer, Consumer}
import java.util.stream.{IntStream, Stream}

import com.google.common.annotations.Beta
import com.google.common.collect.{Iterators, Lists, Maps}
import com.mojang.datafixers.DataFixUtils
import com.mojang.datafixers.util.Pair
import com.mojang.serialization.{DataResult, DynamicOps, MapLike, RecordBuilder}
import io.github.legacy_fabric_community.shadowedconfig.codecs.ops.ExtendedOps
import io.github.legacy_fabric_community.shadowedconfig.nbt.mixin.{CompoundTagAccessor, EndTagAccessor, ListTagAccessor, TagAccessor}
import net.fabricmc.fabric.api.util.NbtType
import net.minecraft.nbt._

//noinspection DuplicatedCode
@Beta
object NbtOps extends ExtendedOps[Tag] {
	// FOR JAVA USE ONLY
	val instance: NbtOps.type = this

	private[nbt] val THE_END_TAG = EndTagAccessor.create

	private def createListTag(b: Byte, c: Byte): Tag = {
		if (matches(b, c, 1.toByte)) return new ByteArrayTag(new Array[Byte](0))
		if (matches(b, c, 3.toByte)) new IntArrayTag(new Array[Int](0))
		else new ListTag
	}

	private def matches(b: Byte, c: Byte, d: Byte) = b == d && (c == d || c == 0)

	private def fill(listTag: ListTag, tag: Tag, tag2: Tag): Unit = {
		tag match {
			case abstractListTag2: ListTag =>
				abstractListTag2.asInstanceOf[ListTagAccessor].getValue.forEach((tag4: Tag) => {
					listTag.add(tag4)
				})
			case _ =>
		}
		listTag.add(tag2)
	}

	private def fillAll(listTag: ListTag, tag: Tag, list: util.List[Tag]): Unit = {
		tag match {
			case tempList: ListTag =>
				tempList.asInstanceOf[ListTagAccessor].getValue.forEach((tag4: Tag) => {
					listTag.add(tag4)
				})
			case _ =>
		}
		list.forEach((tag4: Tag) => {
			listTag.add(tag4)
		})
	}

	override def empty: Tag = NbtOps.THE_END_TAG

	override def convertTo[U](dynamicOps: DynamicOps[U], tag: Tag): U = tag.getType match {
		case NbtType.END =>
			dynamicOps.empty
		case NbtType.BYTE =>
			dynamicOps.createByte(tag.asInstanceOf[Tag.NumberTag].getByte)
		case NbtType.SHORT =>
			dynamicOps.createShort(tag.asInstanceOf[Tag.NumberTag].getShort)
		case NbtType.INT =>
			dynamicOps.createInt(tag.asInstanceOf[Tag.NumberTag].getInt)
		case NbtType.LONG =>
			dynamicOps.createLong(tag.asInstanceOf[Tag.NumberTag].getLong)
		case NbtType.FLOAT =>
			dynamicOps.createFloat(tag.asInstanceOf[Tag.NumberTag].getFloat)
		case NbtType.DOUBLE =>
			dynamicOps.createDouble(tag.asInstanceOf[Tag.NumberTag].getDouble)
		case NbtType.BYTE_ARRAY =>
			dynamicOps.createByteList(ByteBuffer.wrap(tag.asInstanceOf[ByteArrayTag].getArray))
		case NbtType.STRING =>
			dynamicOps.createString(tag.asInstanceOf[TagAccessor].invokeAsString)
		case NbtType.LIST =>
			this.convertList(dynamicOps, tag)
		case NbtType.COMPOUND =>
			this.convertMap(dynamicOps, tag)
		case NbtType.INT_ARRAY =>
			dynamicOps.createIntList(util.Arrays.stream(tag.asInstanceOf[IntArrayTag].getIntArray))
		case _ =>
			throw new IllegalStateException("Unknown tag type: " + tag)
	}

	override def getNumberValue(tag: Tag): DataResult[Number] = if (tag.isInstanceOf[Tag.NumberTag]) DataResult.success(tag.asInstanceOf[NumericTag].getNumber)
	else DataResult.error("Not a number")

	override def createNumeric(number: Number): Tag = if (number.isInstanceOf[Integer] || number.isInstanceOf[BigInteger] || number.isInstanceOf[AtomicInteger]) this.createInt(number.intValue)
	else if (number.isInstanceOf[Float]) this.createFloat(number.floatValue)
	else if (number.isInstanceOf[Byte]) this.createByte(number.byteValue)
	else if (number.isInstanceOf[Long] || number.isInstanceOf[AtomicLong]) this.createLong(number.longValue)
	else if (number.isInstanceOf[Short]) this.createShort(number.shortValue)
	else if (number.isInstanceOf[Double] || number.isInstanceOf[BigDecimal]) this.createDouble(number.doubleValue)
	else throw new UnsupportedOperationException("Only Primitive Numbers are supported!")

	override def createByte(b: Byte) = new ByteTag(b)

	override def createShort(s: Short) = new ShortTag(s)

	override def createInt(i: Int) = new IntTag(i)

	override def createLong(l: Long) = new LongTag(l)

	override def createFloat(f: Float) = new FloatTag(f)

	override def createDouble(d: Double) = new DoubleTag(d)

	override def createBoolean(bl: Boolean) = new ByteTag((if (bl) 1
	else 0).toByte)

	override def getStringValue(tag: Tag): DataResult[String] = if (tag.isInstanceOf[StringTag]) DataResult.success(tag.asInstanceOf[TagAccessor].invokeAsString)
	else DataResult.error("Not a string")

	override def createString(string: String) = new StringTag(string)

	override def mergeToList(tag: Tag, tag2: Tag): DataResult[Tag] = if (!tag.isInstanceOf[ListTag] && !tag.isInstanceOf[EndTag]) DataResult.error("mergeToList called with not a list: " + tag, tag)
	else {
		val listTag = NbtOps.createListTag((tag match {
			case tag1: ListTag => tag1.getElementType
			case _ => 0
		}).toByte, tag2.getType)
		listTag match {
			case tag1: ListTag => NbtOps.fill(tag1, tag, tag2)
			case _ =>
		}
		DataResult.success(listTag)
	}

	override def mergeToList(tag: Tag, list: util.List[Tag]): DataResult[Tag] = if (!tag.isInstanceOf[ListTag] && !tag.isInstanceOf[EndTag]) DataResult.error("mergeToList called with not a list: " + tag, tag)
	else {
		val abstractListTag = NbtOps.createListTag(tag match {
			case tag1: ListTag => tag1.getElementType.toByte
			case _ => 0.toByte
		}, list.stream.findFirst.map((tag: Tag) => {
			tag.getType
		}).orElse(0.toByte))
		abstractListTag match {
			case tag1: ListTag => NbtOps.fillAll(tag1, tag, list)
			case _ =>
		}
		DataResult.success(abstractListTag)
	}

	override def mergeToMap(mapTag: Tag, keyTag: Tag, tag: Tag): DataResult[Tag] = if (!mapTag.isInstanceOf[CompoundTag] && !mapTag.isInstanceOf[EndTag]) DataResult.error("mergeToMap called with not a map: " + mapTag, mapTag)
	else if (!keyTag.isInstanceOf[StringTag]) DataResult.error("key is not a string: " + keyTag, mapTag)
	else {
		val compoundTag = new CompoundTag
		mapTag match {
			case compoundTag2: CompoundTag =>
				compoundTag2.getKeys.forEach((string: String) => compoundTag.put(string, compoundTag2.get(string)))
			case _ =>
		}
		compoundTag.put(keyTag.asInstanceOf[TagAccessor].invokeAsString, tag)
		DataResult.success(compoundTag)
	}

	override def mergeToMap(tag: Tag, mapLike: MapLike[Tag]): DataResult[Tag] = if (!tag.isInstanceOf[CompoundTag] && !tag.isInstanceOf[EndTag]) DataResult.error("mergeToMap called with not a map: " + tag, tag)
	else {
		val compoundTag = new CompoundTag
		tag match {
			case compoundTag2: CompoundTag =>
				compoundTag2.getKeys.forEach((string: String) => compoundTag.put(string, compoundTag2.get(string)))
			case _ =>
		}
		val list : util.ArrayList[Tag] = Lists.newArrayList
		mapLike.entries.forEach((pair: Pair[Tag, Tag]) => {
			def foo(pair: Pair[Tag, Tag]) = {
				val tag2 = pair.getFirst
				if (!tag.isInstanceOf[StringTag]) list.add(tag)
				else compoundTag.put(tag2.asInstanceOf[TagAccessor].invokeAsString, pair.getSecond)
			}

			foo(pair)
		})
		if (list.isEmpty) DataResult.error("some keys are not strings: " + list, compoundTag)
		else DataResult.success(compoundTag)
	}

	override def getMapValues(tag: Tag): DataResult[Stream[Pair[Tag, Tag]]] = if (!tag.isInstanceOf[CompoundTag]) DataResult.error("Not a map: " + tag)
	else {
		val compoundTag = tag.asInstanceOf[CompoundTag]
		DataResult.success(compoundTag.getKeys.stream.map((string: String) => Pair.of(this.createString(string), compoundTag.get(string))))
	}

	override def getMapEntries(tag: Tag): DataResult[Consumer[BiConsumer[Tag, Tag]]] = if (!tag.isInstanceOf[CompoundTag]) DataResult.error("Not a map: " + tag)
	else {
		val compoundTag = tag.asInstanceOf[CompoundTag]
		DataResult.success((consumer: BiConsumer[Tag, Tag]) => compoundTag.getKeys.forEach((string: String) => consumer.accept(this.createString(string), compoundTag.get(string))))
	}

	override def getMap(tag: Tag): DataResult[MapLike[Tag]] = if (!tag.isInstanceOf[CompoundTag]) DataResult.error("Not a map: " + tag)
	else {
		val compoundTag = tag.asInstanceOf[CompoundTag]
		DataResult.success(new MapLike[Tag]() {
			override def get(tag: Tag): Tag = compoundTag.get(tag.asInstanceOf[TagAccessor].invokeAsString)

			override

			def get(string: String): Tag = compoundTag.get(string)

			override

			def entries: Stream[Pair[Tag, Tag]] = compoundTag.getKeys.stream.map((string: String) => Pair.of(NbtOps.this.createString(string), compoundTag.get(string)))

			override

			def toString: String = "MapLike[" + compoundTag + "]"
		})
	}

	override def createMap(stream: Stream[Pair[Tag, Tag]]): Tag = {
		val compoundTag = new CompoundTag
		stream.forEach((pair: Pair[Tag, Tag]) => compoundTag.put(pair.getFirst.asInstanceOf[TagAccessor].invokeAsString, pair.getSecond))
		compoundTag
	}

	override def getStream(tag: Tag): DataResult[Stream[Tag]] = if (tag.isInstanceOf[ListTag]) DataResult.success(tag.asInstanceOf[ListTagAccessor].getValue.stream)
	else DataResult.error("Not a list")

	override def getList(tag: Tag): DataResult[Consumer[Consumer[Tag]]] = tag match {
		case listTag: ListTag =>
			DataResult.success((consumer: Consumer[Tag]) => {
				listTag.asInstanceOf[ListTagAccessor].getValue.forEach(consumer)
			})
		case _ => DataResult.error("Not a list: " + tag)
	}

	override def getByteBuffer(tag: Tag): DataResult[ByteBuffer] = tag match {
		case tag1: ByteArrayTag => DataResult.success(ByteBuffer.wrap(tag1.getArray))
		case _ => super.getByteBuffer(tag)
	}

	override def createByteList(byteBuffer: ByteBuffer) = new ByteArrayTag(DataFixUtils.toArray(byteBuffer))

	override def getIntStream(tag: Tag): DataResult[IntStream] = tag match {
		case tag1: IntArrayTag => DataResult.success(util.Arrays.stream(tag1.getIntArray))
		case _ => super.getIntStream(tag)
	}

	override def createIntList(intStream: IntStream) = new IntArrayTag(intStream.toArray)

	override def createList(stream: Stream[Tag]): Tag = {
		val peekingIterator = Iterators.peekingIterator(stream.iterator)
		if (!peekingIterator.hasNext) new ListTag
		else {
			val tag = peekingIterator.peek
			var list3: util.ArrayList[Tag] = null
			tag match {
				case _: ByteTag =>
					list3 = Lists.newArrayList(Iterators.transform(peekingIterator, (tagx: Tag) => if (tagx != null) tagx.asInstanceOf[ByteTag].getByte
					else 0))
					val arr = new Array[Byte](list3.size)
					for (a <- arr.indices) {
						arr(a) = list3.get(a).asInstanceOf[Byte]
					}
					new ByteArrayTag(arr)
				case _: IntTag =>
					list3 = Lists.newArrayList(Iterators.transform(peekingIterator, (tagx: Tag) => if (tagx != null) tagx.asInstanceOf[IntTag].getInt
					else 0))
					val arr = new Array[Int](list3.size)
					for (a <- arr.indices) {
						arr(a) = list3.get(a).asInstanceOf[Int]
					}
					new IntArrayTag(arr)
				case _ =>
					val listTag = new ListTag
					while ( {
						peekingIterator.hasNext
					}) {
						val tag2 = peekingIterator.next
						if (!tag2.isInstanceOf[EndTag]) listTag.add(tag2)
					}
					listTag
			}
		}
	}

	override def remove(tag: Tag, string: String): Tag = tag match {
		case compoundTag: CompoundTag =>
			val compoundTag2 = new CompoundTag
			compoundTag.getKeys.stream.filter((string2: String) => !Objects.equals(string2, string)).forEach((stringx: String) => compoundTag2.put(stringx, compoundTag.get(stringx)))
			compoundTag2
		case _ => tag
	}

	override def toString = "Nbt"

	override def mapBuilder = new MapBuilder

	private[nbt] class MapBuilder extends RecordBuilder.AbstractStringBuilder[Tag, CompoundTag](NbtOps.this) {
		override protected def initBuilder = new CompoundTag

		override protected def append(string: String, tag: Tag, compoundTag: CompoundTag): CompoundTag = {
			compoundTag.put(string, tag)
			compoundTag
		}

		override protected def build(compoundTag: CompoundTag, tag: Tag): DataResult[Tag] = if (tag != null && (tag ne NbtOps.THE_END_TAG)) if (!tag.isInstanceOf[CompoundTag]) DataResult.error("mergeToMap called with not a map: " + tag, tag)
		else {
			val compoundTag2 = new CompoundTag
			compoundTag.asInstanceOf[CompoundTagAccessor].setData(Maps.newHashMap(tag.asInstanceOf[CompoundTagAccessor].getData))
			compoundTag.asInstanceOf[CompoundTagAccessor].getData.entrySet.forEach((stringTagEntry: util.Map.Entry[String, Tag]) => {
				compoundTag2.put(stringTagEntry.getKey, stringTagEntry.getValue)
			})

			DataResult.success(compoundTag2)
		}
		else DataResult.success(compoundTag)
	}
}
